pallets/parachain-system/src/lib.rs

```diff
@@ -155,6 +155,19 @@ decl_module! {
 			Self::schedule_upgrade_impl(validation_function)?;
 		}
 
+		/// Force an already scheduled validation function upgrade to happen on a particular block.
+		///
+		/// Note that coordinating this block for the upgrade has to happen independently on the relay
+		/// chain and this parachain. Synchronizing the block for the upgrade is sensitive, and this
+		/// bypasses all checks and and normal protocols. Very easy to brick your chain if done wrong.
+		#[weight = (0, DispatchClass::Operational)]
+		pub fn set_upgrade_block(origin, relay_chain_block: RelayChainBlockNumber) {
+			ensure_root(origin)?;
+			if let Some((_, validation_function)) = PendingValidationFunction::get() {
+				PendingValidationFunction::put((relay_chain_block, validation_function));
```

Comment:
Note that this storage pattern is really inefficient

I see in the inherent below that we read this storage item every block just to look at the `relay_chain_block`.

We def want to split this into two storage items or a map so that we don't read the entire validation function every block just to check the upgrade block.

Am I missing anything @pepyakin ?

Reply:
Good point. AFAIK this was not introduced by Sergei. 

But we should fix this.

Reply:
I will add it to this PR